Add NuGet package version to ThisAssembly

I'm working on a .NET tool which will add a reference to a NuGet package to a project file, very similar to what `nbgv` does.

For this, I'd like to use a property on `ThisAssembly` which would return the NuGet package version.

@AArnott  what do you think?

> > > If there was a NuGetPackageVersion available in ThisAssembly, it would be easy to compare.
> > 
> > 
> > Really? What compare method would you use? You'd need to know whether the version string is using semver1 or 2 rules then use a compare function appropriate for that semver syntax. It wouldn't be a simple `Version` class comparer. Would you redist a NuGet assembly and use its comparer function?
> 
> correct, using Nuget.Versioning to provide SemVer-compatible version comparison, to enable auto-update for the application itself, and (future plan) application content, aside using Nuget Core/Protocolto handle feed, download and unpacking. All worked well from the ground for public releases, just got stuck when offering an additional update channel for "developer versions" which are basically non-public releases in nbgv semantic (and served of a separate feed). Certainly not a standard use case.
> 
> > > I've been trying to use the AdditionalThisAssemblyFields
> > 
> > 
> > That would have _very_ specific build ordering requirements. But it should be possible. Like this:
> > ```
> > <Target Name="AddNuGetPackageVersionToThisAssembly"
> >         DependsOnTargets="GetBuildVersion"
> >         BeforeTargets="GenerateAssemblyVersionInfo">
> >   <ItemGroup>
> >     <AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="$(NuGetPackageVersion)" />
> >   </ItemGroup>
> > </Target>
> > ```
> 
> aweseome just did a quick test and seems to do that I needed, saves me from string-mangeling metadata into release labels for non-public releases. Thanks for quick response, and great tool!



> > If there was a NuGetPackageVersion available in ThisAssembly, it would be easy to compare.
> 
> Really? What compare method would you use? You'd need to know whether the version string is using semver1 or 2 rules then use a compare function appropriate for that semver syntax. It wouldn't be a simple `Version` class comparer. Would you redist a NuGet assembly and use its comparer function?
> 
> > I've been trying to use the AdditionalThisAssemblyFields
> 
> That would have _very_ specific build ordering requirements. But it should be possible. Like this:
> 
> ```
> <Target Name="AddNuGetPackageVersionToThisAssembly"
>         DependsOnTargets="GetBuildVersion"
>         BeforeTargets="GenerateAssemblyVersionInfo">
>   <ItemGroup>
>     <AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="$(NuGetPackageVersion)" />
>   </ItemGroup>
> </Target>
> ```

somehow after 3.4.255 this does not seem to work anymore, in 3.5.103+ it the additional field is not created anymore, breaking our code (error CS0117: 'ThisAssembly' does not contain a definition for 'NuGetPackageVersion'). Any idea @AArnott ?

--- a/doc/msbuild.md
+++ b/doc/msbuild.md
@@ -38,6 +38,47 @@ You should only build with this property set from one release branch per
 major.minor version to avoid the risk of producing multiple unique NuGet
 packages with a colliding version spec.
 
+## Customization through MSBuild
+
+### Properties
+
+The following MSBuild properties may be set in your project to customize certain Nerdbank.GitVersioning behaviors:
+
+Property | Default | Description
+--|--|--
+`NBGV_ThisAssemblyNamespace` | `$(RootNamespace)` | Sets the namespace to use for the generated `ThisAssembly` class.
+`NBGV_EmitThisAssemblyClass` | `true` | When `false`, suppresses generation of the `ThisAssembly` class.
+`NBGV_ThisAssemblyIncludesPackageVersion` | `false` | When `true`, a `NuGetPackageVersion` property is added to the `ThisAssembly` class.
+
+### Custom `ThisAssembly` static fields and constants
+
+Custom constants may be added to the `ThisAssembly` class through `AdditionalThisAssemblyFields` items defined in your project.
+The item name becomes the constant's name. The value of the constant may be defined by metadata whose name determines the type of the constant as follows:
+
+Metadata name | Property type | Notes
+--|--|--
+`String` | `string` | These may also carry `EmitIfEmpty="true"` metadata to force generation of the constant even if the string value is empty.
+`Boolean` | `bool` |
+`Ticks` | `DateTime` | The integer specified in msbuild metadata will be interpreted as ticks to construct the `DateTime` value.
+
+For example, to the following MSBuild items:
+
+```xml
+<ItemGroup>
+  <AdditionalThisAssemblyFields Include="FinalRelease" Boolean="true" />
+  <AdditionalThisAssemblyFields Include="MyMessage" String="Hello World" />
+  <AdditionalThisAssemblyFields Include="AGoodMoment" Ticks="638300335845163552" />
+</ItemGroup>
+```
+
+Will add the following constants or static fields to the `ThisAssembly` class:
+
+```cs
+internal const bool FinalRelease = true;
+internal const string MyMessage = "Hello World";
+internal static readonly DateTime AGoodMoment = new System.DateTime(638300335845163552L, System.DateTimeKind.Utc);
+```
+
 ## Custom build authoring
 
 If you are writing your own MSBuild targets or properties and need to consume version information,

--- a/src/Nerdbank.GitVersioning.Tasks/build/Nerdbank.GitVersioning.targets
+++ b/src/Nerdbank.GitVersioning.Tasks/build/Nerdbank.GitVersioning.targets
@@ -120,6 +120,9 @@
       <VersionSourceFile>$([MSBuild]::NormalizePath('$(IntermediateOutputPath)', '$(AssemblyName).Version$(DefaultLanguageSourceExtension)'))</VersionSourceFile>
       <NewVersionSourceFile>$(VersionSourceFile).new</NewVersionSourceFile>
     </PropertyGroup>
+    <ItemGroup>
+      <AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="$(NuGetPackageVersion)" Condition="'$(NBGV_ThisAssemblyIncludesPackageVersion)' == 'true'" />
+    </ItemGroup>
     <MakeDir Directories="$(IntermediatePath)"/>
     <Nerdbank.GitVersioning.Tasks.AssemblyVersionInfo
                   OutputFile="$(NewVersionSourceFile)"

